---
title: "Differential Expression and Heatmaps"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r}
#install.packages("BiocManager")
#BiocManager::install("Seurat")
#BiocManager::install("ggplot2")
#install.packages("dplr")
#install.packages("ape")
#install.packages("cowplot")
#install.packages("Matrix")
```

Load packages into your workspace. 

```{r}
library(Seurat)
library(ggplot2)
library(dplyr)
library(ape)
library(cowplot)
library(Matrix)

#find out where you are
getwd()
```

*Edit the following code every time* 
and make sure the folders "QC" and ... are present in the "dir" folder
```{r}
#Specify your preferred directory for all input + output
dir= "/Users/whippoorwill/Desktop/Sequencing/LD_AVM02"

#Specify exactly where your seurat files live
datafolder = "Data/Seurat"

#Specify your organism; please capitalize the first letter (i.e. "Mouse", "Human","Zebrafish","Rat")
organism = "Mouse"

#metrics you want to look at for QC
m = c("nCount_RNA","nFeature_RNA","percent.mito","percent.ribo")

#set up folders
QCfolder = "QC"

#This name needs to match your project name within the seurat object
project<-"Microglia_BC" 

#You can add in housekeeping genes if you want them to be scaled always; otherwise set to "NULL"
add.genes = NULL

#Choose what to regress out - could be age, sex, or any metadata column
regress = c("nFeature_RNA")

#Decide whether to use SCtransform (can take a long time with big datasets; generally recommended)
sct = TRUE

#How many genes do you want scaled to use for UMAP/clustering? 
ngenes = 10000

#Which principal components do you want to calculate on? This is a default setting, change if one of the pc's is for something you think is a technical error (i.e. HSP, RP, etc)
pcs = 1:30

#clustering resolution; the last number will be saved as "seurat_clusters" in metadata
res = c(0.5,1.5,1.0)

#Important genes to determine your cells of interest
igenes = c("Cx3cr1","Rbfox3","Aldh1l1","Pecam1")

#metadata dimensions
dims = c("seurat_clusters","sample_description","age","condition")

#the data file you want to open
filename = "_Macrophages_subset.RData"
```

Load in your filtered dataset
```{r}
load(file.path(dir,datafolder,paste0(project,filename)))
```


Heatmap for clusters within this celltype
```{r}
Idents(sobject) = "seurat_clusters"
markers_all <- FindAllMarkers(
    object = sobject,
    features = rownames(sobject),
    test.use = "MAST",
    only.pos = FALSE, 
    min.pct = 0.05, #gene must be present in 5% of the cells in the cluster
    logfc.threshold = 0.0
)

write.csv(markers_all,file = file.path(dir,paste0(keep,"_all_markers.csv")))
```


```{r}
#set your p-value threshold
pcut = 1e-5
genespercluster = 5

#select positive markers for heatmap
markers_all = markers_all[markers_all$avg_logFC>0.2,]

#select genes that pass a p-value threshold
markers_all = markers_all[markers_all$p_val_adj<pcut,]
markers_all <- markers_all[markers_all$gene %in% names(table(markers_all$gene))[table(markers_all$gene) == 1],] #markers that define a single cluster

topgenes <- markers_all %>% group_by(cluster) %>% top_n(genespercluster, avg_logFC)

setEPS()
postscript(file.path(dir,paste0("heatmap_",keep,"_only",pcut,".eps")))
DoHeatmap(
    object = sobject, 
    features = topgenes$gene,
    group.by = 'seurat_clusters',
    size = 5,
    label = T,
    draw.lines = T
)

dev.off()
```
Heatmap for console

```{r}
DoHeatmap(
    object = sobject, 
    features = topgenes$gene,
    group.by = 'seurat_clusters',
    size = 5,
    label = T,
    draw.lines = T
)
```


```{r}
save(sobject,file = file.path(dir,paste0(keep,"_only.RData")))
```

As this is a checkpoint, you can load in the data again if necessary: 

```{r}
load(file.path(dir,"mpeg_only_sctransform.RData"))
de = read.csv(file.path(dir,"all_markers_mpeg.csv"),stringsAsFactors = F)
head(de)
```

```{r}
genes = c("pcna","siglec15l")
DimPlot(sobject,label = T)
DimPlot(sobject,group.by = "sample_description")

FeaturePlot(sobject,genes)
VlnPlot(sobject,genes,pt.size = 0.001,sort = T)

sobject = BuildClusterTree(sobject,dims = 1:30)
PlotClusterTree(sobject)
```


```{r}
table(sobject$age,sobject$seurat_clusters)
```


```{r}
type1 = c(0,2,5,8)
name1 = "juvenile"
type2 = c(1,4,7)
name2 = "adult"
type3 = c(3)
name3 = "macrophage"
type4 = 6
name4 = "proliferating"

#Initialize the cluster levels as a vector and replace the cluster levels with the appropriate name. 
clusters = as.factor(sobject$seurat_clusters)
type = levels(clusters)
type[type1+1] = name1
type[type2+1] = name2
type[type3+1] = name3
type[type4+1] = name4
levels(clusters) = type

#Add a metadata column
sobject$celltype = clusters
sobject$celltypecluster = paste0(sobject$celltype,"-",sobject$seurat_clusters)

DimPlot(sobject,group.by = "celltypecluster",split.by = "age")
```

Block to print multiple graphs: 
```{r}
name = "all_mpeg"
genes = c("marco","mpeg1.1","apoeb","p2ry12","ptprc","hexb","irf8","spi1b","spi1a","csf1ra","csf1rb","slc7a7","c1qa","c1qb","c1qc","cd74a","cd74b","pcna","mki67","siglec15l")
features = c("percent.mito","percent.ribo","nCount_RNA","nFeature_RNA")
groups = c("age","sample_description","seurat_clusters","celltype","celltypeclusters")
genes = genes[genes %in% rownames(GetAssayData(sobject,slot = "data"))]

#just to print to the console some nice graphs
FeaturePlot(sobject,genes[1:5])
FeaturePlot(sobject,genes[6:9])

#feature plots
for(feature in c(genes,features)){
  PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
}

#split feature plots by individual
for(feature in c(features)){
  PrintSeuratGraph(namecard = paste0(name,"_split"),graphtype = "feature",feature = feature,split = "sample_description")
}

#dim plots for clustering
for(group in groups){
  PrintSeuratGraph(namecard = name,graphtype = "dim",group = group, feature = group)
}

#violin plots
for(feature in c(genes,features)){
  PrintSeuratGraph(namecard = name,graphtype = "violin",feature = feature,group = "seurat_clusters")
}
```

```{r}
Idents(sobject) = "celltypecluster"
sobject = BuildClusterTree(sobject,dims = c(1:30))
tree = sobject@tools$BuildClusterTree
setEPS()
postscript(file.path(dir,paste0(name,"tree.eps")))
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
dev.off()
```

Conduct differential expression analysis on the new metadata
```{r}
Idents(sobject) = "celltype"
markers_all = FindAllMarkers(
  object = sobject,
  features = rownames(sobject),
  test.use = "MAST", 
  only.pos = FALSE, 
  min.pct = 0.10, 
  logfc.threshold = 0.0)

```

```{r}
filename = "juv_v_adult_mpeg"
write.csv(markers_all,file = file.path(dir, paste0(filename,".csv")))

single = TRUE
pcut = 1e-25
lfc = 0.2
ngenes = 5

markers_all = markers_all[markers_all$p_val_adj<pcut,]
markers_all = markers_all[markers_all$avg_logFC>lfc,]

#markers that define a single cluster
if (single){
  markers_all_single = markers_all[
    markers_all$gene %in% names(table(markers_all$gene))[table(markers_all$gene) == 1],]
} 


topgenes <- markers_all %>% group_by(cluster) %>% top_n(ngenes, avg_logFC)

setEPS()
postscript(file.path(dir,paste0("heatmap_",filename,".eps")))
DoHeatmap(
    object = sobject, 
    features = topgenes$gene,
    group.by = 'celltypecluster',
    size = 5,
    label = T,
    draw.lines = T
)

dev.off()

```

Heatmap for console
```{r}
DoHeatmap(
    object = sobject, 
    features = topgenes$gene,
    group.by = 'celltypecluster',
    size = 5,
    label = T,
    draw.lines = T
)
```

