---
title: "Atac-seq"
output: html_notebook
---
```{r}
library(Seurat)
library(dplyr)
```


Load file with differentially expressed genes

```{r}
#if public: 
#genes = read.csv("https://github.com/lcdorman/scrnaseq/blob/master/allmarkers_vargenesMG_0501.csv",stringsAsFactors = F)

#if private:
genes = read.csv("allmarkers_vargenesMG_0501.csv",stringsAsFactors = F)

#remove excess columns
genes = genes[,3:ncol(genes)]
head(genes)

#remove genes that don't pass a threshold for p-value and log fold change
alpha = 1e-25
lfoldchange = 0.2

genes = genes[genes$p_val_adj<alpha,]
genes = genes[genes$avg_logFC > lfoldchange | genes$avg_logFC < -lfoldchange,]
```

Load file with atac-seq peaks and edit to merge each gene into one line

```{r}
#if public: 
#peaks = read.csv("https://github.com/lcdorman/scrnaseq/blob/master/MergedpeaksPBSAnnotated_v2.csv",stringsAsFactors = F)

#if private:
peaks = read.csv("MergedpeaksPBSAnnotated_v2.csv",stringsAsFactors = F)
head(peaks)
peaks = peaks[,c("Peak.Score","Focus.Ratio.Region.Size","Detailed.Annotation","Gene.Name")]
head(peaks)
```

Change "Detailed.Annotation" to specific meaningful values
```{r}
annotation = peaks$Detailed.Annotation
rpt = grep("repeat",annotation)
names(rpt) = rep("Repeat",length(rpt))

#pull out different kinds of annotations
prom = grep("promoter",annotation)
names(prom) = rep("Promoter",length(prom))

intergenic = grep("Intergenic",annotation)
names(intergenic) = rep("Intergenic",length(intergenic))

cpg = grep("CpG",annotation)
names(cpg) = rep("CpG",length(cpg))

utr3 = grep("3' UTR",annotation)
names(utr3) = rep("UTR3",length(utr3))

utr5 = grep("5' UTR",annotation)
names(utr5) = rep("UTR5",length(utr5))

exon = grep("exon",annotation)
names(exon) = rep("Exon",length(exon))
exon = exon[!exon %in% c(utr3,utr5)]

intron = grep("intron",annotation)
names(intron) = rep("Intron",length(intron))

tts = grep("TTS",annotation)
names(tts) = rep("TTS",length(tts))

annotationfull = c(rpt,exon,prom,intergenic,cpg,utr3,utr5,intron,tts)
annotationfull = annotationfull[order(annotationfull)]

annotation[annotationfull] = names(annotationfull)
```

Annotate peaks
```{r}
peaks$Detailed.Annotation = annotation

#Separate only the genes that are present in my file
allgenes = genes$gene
peaks = peaks[peaks$Gene.Name %in% allgenes,]

#Include only confident peaks (> 1 sample)
peaks = peaks[grep("\\|",peaks$Focus.Ratio.Region.Size),]
head(peaks)
```

Select only the genes with promoter peaks (skip or change "Promoter" based on what you are interested in looking at)
```{r}
peaks = peaks[peaks$Detailed.Annotation == "Promoter",]
peaks = peaks[order(peaks$Peak.Score,decreasing = T),]
head(peaks)
```

Merge every annotation for the same gene into one line. 
```{r}
k = ncol(peaks)
for (i in 2:length(peaks$Gene.Name)){
  if (peaks$Gene.Name[(i-1)] == peaks$Gene.Name[i]){ 
    peaks[(i-(k/4)),(k+1):(k+4)] = peaks[i,1:4]
    k = k+4
  } else {k = 4}
}

peaks = peaks[!duplicated(peaks$Gene.Name),]
head(peaks)
```

Select only the annotations for genes in your file, and save for future reference
```{r}
peaks = peaks[match(genes$gene,peaks$Gene.Name),]
atac =cbind(genes,peaks)
write.csv(atac,file = "atacseq_degenes_promoters.csv")
```

(If starting in the middle, load the previously saved file)
```{r}
atac = read.csv("atacseq_degenes_promoters.csv",stringsAsFactors = F)
```

Make a new spreadsheet with only the genes upregulated in a specific cluster
```{r}
cluster = "3"
genes_cluster = genes[genes$cluster == cluster,]
dim(genes_cluster)
peaks_cluster = peaks[peaks$Gene.Name %in% genes_cluster$gene,]
dim(peaks_cluster)
head(peaks_cluster)
```

Write a spreadsheet (note: 315 genes in cluster 3, 136 peaks)

```{r}
write.csv(peaks_cluster,paste0("Peaks_cluster",cluster,".csv"))
```

Compare this to, say, cluster 4: 

Make a new spreadsheet with only the genes upregulated in cluster 4

```{r}
cluster = "4"
genes_cluster = genes[genes$cluster == cluster,]
peaks_cluster = peaks[peaks$Gene.Name %in% genes_cluster$gene,]
write.csv(peaks_cluster,paste0("Peaks_cluster",cluster,".csv"))
```


Compare this to, say, cluster 8: 

Make a new spreadsheet with only the genes upregulated in cluster 8
```{r}
cluster = "8"
genes_cluster = genes[genes$cluster == cluster,]
peaks_cluster = peaks[peaks$Gene.Name %in% genes_cluster$gene,]
write.csv(peaks_cluster,paste0("Peaks_cluster",cluster,".csv"))
```

Compare with barres dataset: (Courtesy of http://brainrnaseq.org)

```{r}
#private
barres = read.csv("barreslab_rnaseq.csv",stringsAsFactors = F)

#public
#barres = read.csv("https://github.com/lcdorman/scrnaseq/blob/master/barreslab_rnaseq.csv",stringsAsFactors = F)

```

Calculate a cell type specificity rating for your cell type of interest (in this case, microglia)

```{r}
head(barres)
colnames(barres)
```
Specify which columns are your cell of interest (coi) and which are other cells (other)

```{r}
coi = 8
other = c(3:(coi-1),(coi+1):9)
```

This function will calculate a Cell Type specificity by dividing the fpkm for your cell type by the average fpkm for all other cell types.
```{r}
for (i in 1:nrow(barres)){
  row = barres[i,]
  mspec = row[coi]/mean(row[other],na.rm = T)
  barres[i,10] = mspec
}
colnames(barres)[10] = paste0(colnames(barres)[coi],"_Specificity")

```

```{r}
barres2 = barres[match(atac$Gene,barres$Gene.symbol),]
head(atac)
head(barres2)
dim(atac)
dim(barres2)
atac2 = cbind(barres2,atac)
```


```{r}
write.csv(atac2,"~/Desktop/Atac-seq/atacseqwithbarres_promoters.csv")
```

```{r}
b3 = barres[barres$Gene.symbol %in% atac_cluster,]
```

Relationship between peak size and microglial specificity by cluster

```{r}
atac = read.csv("~/Desktop/Sequencing/LD_AVM02/Atac-seq/atacseqwithbarres_promoters.csv",stringsAsFactors = F)
a = cbind("Gene" = atac$Gene.symbol,"Cluster" = atac$cluster,"Specificity" = atac$MicrogliaSpecificity,"Peak" = atac$Peak.Score,atac$Peak.Score.1,atac$Peak.Score.2,atac$Peak.Score.3)
head(a)
for (i in 1:nrow(a)){
  peak = max(a[i,4:7],na.rm = T)
  a[i,4] = peak
}
head(a)
a = a[,1:4]
head(a)
class(a)
a = as.data.frame(a)
head(a)
head(a$Peak)
a$Peak = as.character(a$Peak)
a$Peak = as.numeric(a$Peak)

a$Specificity = as.character(a$Specificity)
a$Specificity = as.numeric(a$Specificity)
head(a)


library(ggplot2)
theme_set(
  theme_bw() +
    theme(legend.position = "top")
  )


b <- ggplot(a[a$Cluster %in% c(0,3,4,8),], aes(x = log(Specificity), y = Peak))
setEPS()
postscript("~/Desktop/atacseq.eps")
b + geom_point(aes(color = Cluster)) +
  scale_color_manual(values = c('darkgrey','green','blue','purple'))
dev.off()


b <- ggplot(a[a$Cluster %in% c(0),], aes(x = log(Specificity), y = Peak))
setEPS()
postscript("~/Desktop/atacseq.eps")
b + geom_point(aes(color = Cluster)) +
  scale_color_manual(values = c('darkgrey','green','blue','purple'))
dev.off()


b <- ggplot(a[a$Cluster %in% c(3),], aes(x = log(Specificity), y = Peak))
setEPS()
postscript("~/Desktop/atacseq3.eps")
b + geom_point(aes(color = Cluster)) +
  scale_color_manual(values = c('green','blue','purple'))
dev.off()

b <- ggplot(a[a$Cluster %in% c(4),], aes(x = log(Specificity), y = Peak))
setEPS()
postscript("~/Desktop/atacseq4.eps")
b + geom_point(aes(color = Cluster)) +
  scale_color_manual(values = c('blue','purple'))
dev.off()

b <- ggplot(a[a$Cluster %in% c(8),], aes(x = log(Specificity), y = Peak))
setEPS()
postscript("~/Desktop/atacseq8.eps")
b + geom_point(aes(color = Cluster)) +
  scale_color_manual(values = c('purple'))
dev.off()
```

adding in spliced:unspliced ratio

```{r}
ratio = read.csv("/Users/whippoorwill/Desktop/Jupyter/spliced_unspliced_ratio.csv",stringsAsFactors = F,header = F)
ratio = as.numeric(ratio)
genes = read.csv("/Users/whippoorwill/Desktop/Jupyter/colnames.csv",stringsAsFactors = F)
names(ratio) = genes$Name

```

add ratios to the spreadsheet a
```{r}
ratio = ratio[match(a$Gene,names(ratio))]
a$splicedratio = ratio
```
Peak chromatin score vs splicedratio

```{r}
ac = a[a$Cluster %in% c(0,3,4,8),]

b <- ggplot(ac, aes(x = log(splicedratio), y = Peak))


c = b + geom_point(aes(color = Cluster)) +
  scale_color_manual(values = c('darkgrey','green','blue','purple')) + geom_text( 
    data=ac %>% filter(log(splicedratio)>6 | log(splicedratio)<(-3) | Peak>120),
    aes(label=Gene))   
c
setEPS()
postscript("~/Desktop/atacseqpeak_splicedratio.eps")
c
dev.off()
```

Peak chromatin score vs specificity

```{r}
ac = a[a$Cluster %in% c(3),]

b <- ggplot(ac, aes(y = Peak, x = log(Specificity)))


c = b + geom_point(aes(color = Cluster)) +
  scale_color_manual(values = c('darkgrey','green','blue','purple')) + geom_text( 
    data=ac %>% filter(log(splicedratio)>6 | log(splicedratio)<(-3) | log(Specificity)>5),
    aes(label=Gene))   
c
setEPS()
postscript("~/Desktop/specificity_peak.eps")
c
dev.off()
```


```{r}
ac = a[a$Cluster %in% c(0,3,4,8),]

b <- ggplot(ac, aes(x = log(splicedratio), y = log(Specificity)))


c = b + geom_point(aes(color = Cluster)) +
  scale_color_manual(values = c('darkgrey','green','blue','purple')) + geom_text( 
    data=ac %>% filter(log(splicedratio)>6 | log(splicedratio)<(-3) | log(Specificity)>5),
    aes(label=Gene))   
c
setEPS()
postscript("~/Desktop/specificity_splicedratio.eps")
c
dev.off()
```
Volcano Plot: 



```{r}
#fc = markers_all
fc = read.csv("/Users/whippoorwill/Dropbox (Anna Molofsky Lab)/2020-Leah-barrelCortex/Manuscript data/Glia RNASeq/P5_P7 MG rnaseq/Spreadsheets/allmarkers_vargenesMG_0501.csv",stringsAsFactors = F)
colnames(fc)[8] = "Gene"
```

```{r}
newlist = list()
#Split by cluster
for (i in c(0:9)){
  newlist[[i+1]] = fc[fc$cluster == i,]
}
cluster = 3

#select a single cluster
setwd("~/Desktop/plots")
fc = newlist[[cluster+1]]
fc = fc[!is.na(fc$avg_logFC),]
colorkeysdown = fc$Gene[fc$avg_logFC < -log2(1.15) & fc$p_val_adj < 10e-25]
colorkeysup = fc$Gene[fc$avg_logFC > log2(1.15) & fc$p_val_adj < 10e-25]
```

Cluster 8: 72 atac+ out of 91
Cluster 4: 208 out of 226
cluster 3: 141 out of 314
cluster 0: 10 of 10 

```{r}
atac = read.csv("/Users/whippoorwill/Desktop/Sequencing/LD_AVM02/Atac-seq/atacseqwithbarres_promoters.csv",stringsAsFactors = F)
atac = atac[atac$Peak.Score>5,]

allcolors = rep("darkgrey",length(fc$Gene))
names(allcolors) = fc$Gene
a = c("Meg3","Thsd7a","Mast1","Sparcl1","Vcan","Adam11","Gria1","Meg3","Nrxn1","Shank1","Ahi1","Ank2","Nrxn2","Ptprs","Cux2","Sema6c","Gabbr1")
#a = c("Casp8ap2","Vim","P2ry12","Mrc1","Cd68","Lgals3","Mki67","Itgax","Gpnmb","Lmnb1","Cd9","Tmem119","Ctsl","Igf1","Apoe","Spp1")
#a = c("Ifit2","Parp12","Axl","Ccl12","Ly6e","Mx1","Isg15","Ifitm3","Lgals9","Irf7","Bst2")
a = a[a %in% atac$Gene]

allcolors[names(allcolors) %in% colorkeysdown] = "blue"
allcolors[names(allcolors) %in% colorkeysup]= "yellow"
allcolors[names(allcolors)%in% atac$Gene] = "red"


names(allcolors)[allcolors == "yellow"] = "u"
names(allcolors)[allcolors == "darkgrey"] = "-"
names(allcolors)[allcolors == "blue"] = "d"
names(allcolors)[allcolors == "red"] = "a"

```

Make the plot
  
```{r}
setEPS()
postscript(paste0("~/Desktop/volcano_atac",cluster,"label2.eps"))
EnhancedVolcano(fc,
                lab = fc$Gene,
                x = 'avg_logFC',
                y = 'p_val_adj',
                xlim = c(-2.5, 2.5),
                title = paste0("LDAVM02_atac",cluster),
                subtitle = "",
                drawConnectors = T,
                legendPosition = 'right',
                legendVisible = F,
                pCutoff = 10e-25,
                FCcutoff = log2(1.15),
                selectLab = a,
                transcriptPointSize = 3,
                transcriptLabSize = 5,
                col=c('black', 'black', 'black', 'red3'),
                colCustom = allcolors,
                gridlines.major = F,
                gridlines.minor = F,
                colAlpha = 1)

dev.off()

```

Make a table
```{r}
fc = read.csv("/Users/whippoorwill/Dropbox (Anna Molofsky Lab)/2020-Leah-barrelCortex/Manuscript data/Glia RNASeq/P5_P7 MG rnaseq/Spreadsheets/allmarkers_vargenesMG_0501.csv",stringsAsFactors = F)
colnames(fc)[8] = "Gene"


```

Split fc by p-value, log fc
```{r}
pval = 10e-25
lfc = 0.2
fc = fc[fc$p_val_adj < pval,]
fcup = fc[fc$avg_logFC > lfc,]
fcdown = fc[fc$avg_logFC < -lfc,]

table(fcup$cluster)
table(fcdown$cluster)
```
compare to atac-seq data
```{r}
atac = read.csv("/Users/whippoorwill/Desktop/Sequencing/LD_AVM02/Atac-seq/atacseqwithbarres_promoters.csv",stringsAsFactors = F)
atac = atac[!is.na(atac$Gene),]
atacup = atac[match(fcup$Gene,atac$Gene),c("Gene","Peak.Score")]
fcup$atac = atacup$Peak.Score
fcup$truefalse = !is.na(fcup$atac)

atacdown = atac[match(fcdown$Gene,atac$Gene),c("Gene","Peak.Score")]
fcdown$atac = atacdown$Peak.Score
fcdown$truefalse = !is.na(fcdown$atac)

"upregulated, p<10e-25,lfc>0.2, true = atacseq >5"
x = table(fcup$cluster,fcup$truefalse)
x = cbind(x,"Total" = rowSums(x))
x = cbind(x,"percent atac>5" = (x[,"TRUE"]/x[,"Total"] * 100))
up = x
up
"downregulated, p<10e-25, lfc < -0.2, true = atacseq > 5"
x = table(fcdown$cluster,fcdown$truefalse)
x = cbind(x,"Total" = rowSums(x))
x = cbind(x,"percent atac>5" = (x[,"TRUE"]/x[,"Total"] * 100))
x = rbind(x[1:2,],"3" = c(0,0,0,0),x[3:7,])
down = x
down
```
Tornado chart

```{r}

barplot(up[,4], horiz = T, las=1, xlim = c(-100,100), xaxt='n', ylab = 'cluster',
        beside=T, col=c('red'))
barplot(down[,4], horiz = T, las=1, xlim = c(-100,100), xaxt='n', ylab = 'cluster',
        beside=T, col=c('gold'), add = TRUE)
axis(1, at=pretty(x),  lab=paste0(pretty(x)," %"), las=TRUE)
```
Bar chart

```{r}

setEPS()
postscript("~/Desktop/barplot_atac_updown.eps")
barplot(up[,4], horiz = F, las=1, xlim = c(0,10),xlab = "cluster", xaxt='n', ylab = '% atac peak > 5',
        beside=F, col=c('red'),ylim = c(0,100))
barplot(down[,4], horiz = F, las=1, xlim = c(0,10), xlab = "cluster", xaxt='n', ylab = '% atac peak > 5',
        beside=F, col=c('gold'), add = TRUE,ylim = c(0,100))

dev.off()


```

cellphonedb
```{r}
cellphonedb = read.csv("/Users/whippoorwill/Desktop/Cellphonedb/significant_means.csv",stringsAsFactors = F)
colnames(cellphonedb)
```
count per interaction, remove rows with nothing
```{r}
cdb = cellphonedb[,13:28]
cdb = cdb[rowSums(cdb,na.rm = T)>0,]
```

Replace NA with 0 and any number with 1 (count incidents)

```{r}
cdb[!is.na(cdb)] = 1
cdb[is.na(cdb)] = 0


cd = colSums(cdb,na.rm = T)
cd
```
Make a table

```{r}
class(cd)
cd = as.data.frame(cd)
dim(cd)
cd$three = cd[1:4,1]
cd$four = cd[5:8,1]
cd$zero = cd[9:12,1]
cd$eight = cd[13:16,1]
```
Make a table of rows = target (3,4,5,8) and columns = source (3,4,5,8)
```{r}
dim(cd)
cd = cd[1:4,2:5]
cd
```
Rename
```{r}
rownames(cd) = c("T3","T4","T0","T8")
colnames(cd) = c("S3","S4","S0","S8")
```
show
```{r}
cd = cd[c(3,1,2,4),c(3,1,2,4)]
cd
```
Make a table

```{r}
heatmap(as.matrix(cd),Rowv = NA,Colv = NA,symm = F,xlab = "Source",ylab = "Target",scale = "none")
```
```{r}
library(pheatmap)
pheatmap(as.matrix(cd), display_numbers = T,scale = "none",cluster_cols = F,cluster_rows = F )
cd = as.matrix(cd)
cd = round(cd,0)

setEPS()
postscript("~/Desktop/heatmap_cellphonedb_mg.eps")
pheatmap(cd, display_numbers = T,scale = "none",cluster_cols = F,cluster_rows = F,color = c("lightgrey","lightyellow","gold","orange","darkorange","red","darkred" ))
dev.off()
```


```{r}
x = atac$Detailed.Annotation[atac$Detailed.Annotation == "Promoter"]
length(x)
```

```{r}
x1 = atac$Detailed.Annotation.1[atac$Detailed.Annotation.1 == "Promoter"]
length(x1)
```

```{r}
x2 = atac$Detailed.Annotation.2[atac$Detailed.Annotation.2 == "Promoter"]
length(x2)
```


```{r}
x3 = atac$Detailed.Annotation.3[atac$Detailed.Annotation.3 == "Promoter"]
length(x3)
```


