---
title: "CellphoneDB-multipledatasets"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

For cellphonedb: need to make "meta.txt" and "counts.txt", put in opc folder in jupyter
should include all microglia clusters + opc clusters + oligo for good measure
```{r}
load("/Users/whippoorwill/Desktop/Sequencing/LD_RNASeq2018/RData_v3/Current_LD1_march30.RData")
```


```{r}
sobject1 = mgAVM02
sobject2 = LD1
sobject2 = subset(sobject2,subset = celltype %in% c("opcs","oligos"))
column = "celltypecluster"
```

Add in a metadata column with celltypecluster
```{r}
meta1 = paste0("microglia_",sobject1$finalclusters)
names(meta1) = colnames(sobject1)
meta2 = sobject2$celltypecluster
meta1 = meta1[meta1 %in% c("microglia_4","microglia_5")]
meta = as.data.frame(c(meta1,meta2))
meta[,1] = as.character(meta[,1])
colnames(meta) = "Cluster"

#meta = cbind("Cellname" = rownames(meta),"Cluster" = meta[,1])
write.table(meta,file = "~/Desktop/Jupyter/OPC/meta.txt",sep = "\t",row.names = T,quote = F)
```

Make a counts file with both opcs and microglia
```{r}
countsm = GetAssayData(sobject1,slot = "counts")
countso = GetAssayData(sobject2,slot = "counts")

o = rownames(countso)
m = rownames(countsm)

length(o)
o =  o[o %in% m]
length(o)

m = m[m %in% o]
length(m)

countsm = countsm[rownames(countsm) %in% m,colnames(countsm) %in% names(meta1)]
countso = countso[rownames(countso) %in% o,]

head(countsm[,1])
head(countso[,1])

countsm = countsm[order(rownames(countsm)),]
countso = countso[order(rownames(countso)),]
```


```{r}
if (!identical(rownames(countso),rownames(countsm))){print("fix gene order")}
if (identical(rownames(countso),rownames(countsm))){
  counts = cbind(countsm,countso)
  rownames(counts) = toupper(rownames(counts)) #convert to ensemble ids
  library(org.Hs.eg.db)
 #any spreadsheet with gene symbols or other identifiers
  ids=rownames(counts) 
  fromKey="SYMBOL" #must match the ids - could also be ensembl ID
  toKey=c("ENSEMBL") #whatever annotation you want to add - find with keytypes(db)
  db=org.Hs.eg.db
  selRes<-AnnotationDbi::select(db,keys=ids,keytype=fromKey,columns=c(fromKey,toKey))

  ##this is really used to get particular keys of the keytype "ensembl" (or any other type)

  x=selRes[match(ids,selRes[,1]),1:2]
  identical(x$SYMBOL,rownames(counts))
  rownames(counts) = x$ENSEMBL
  ids = rownames(counts)
  keep = !is.na(ids)
  counts = counts[keep,]
  write.table(counts,file = "~/Desktop/Jupyter/OPC/counts.txt",sep = "\t", quote = F,row.names = T)
}


```


Cellphone db: also from jupyter, see python script for details. "X3.04" means from cluster 3 (source) to cluster 4 (target)
```{r}
cellphonedb = read.csv("../Spreadsheets/OPC_2significant_means.csv",stringsAsFactors = F,fill = T)
#cellphonedb = read.csv("https://github.com/lcdorman/scrnaseq/blob/master/significant_means.csv",stringsAsFactors = F)
colnames(cellphonedb)
x = grep("opcs",colnames(cellphonedb))
y = grep("oligos",colnames(cellphonedb))

cellphonedb = cellphonedb[,c(1:12,x,y)]
x = grep("microglia_4",colnames(cellphonedb))
y = grep("microglia_5",colnames(cellphonedb))

cellphonedb = cellphonedb[,c(1:12,x,y)]

cdb = cellphonedb
```
count per interaction, remove rows with nothing
```{r}
cdb = cellphonedb[,13:ncol(cellphonedb)]
cdb = cdb[rowSums(cdb,na.rm = T)>0,]
```

Replace NA with 0 and any number with 1 (count incidents)
```{r}
cdb[!is.na(cdb)] = 1
cdb[is.na(cdb)] = 0
cd = colSums(cdb,na.rm = T)
cd
```
**Make a table(* 

First we will repeat the names for source and target and then rename source and target to the appropriate portions of the name itself. 

```{r}
cd = rbind("Source" = names(cd),"Target" = names(cd),"interactions" = cd)
cd["Source",] = sapply(cd["Source",],function(x)paste0("Source_",strsplit(x,"\\.")[[1]][1]))
cd["Target",] = sapply(cd["Target",],function(x)strsplit(x,"\\.")[[1]][2]) #split on ".", take 2nd
#cd["Target",] =sapply(cd["Target",],function(x)paste0("Target_",strsplit(x,"")[[1]][2]))
cd["Target",] =sapply(cd["Target",],function(x)paste0("Target_",x))#remove "0"
```

Make a 3D table first, then iterate through each matrix and add the 3rd dimension (#interactions) to its appropriate position

```{r}
x = table(cd["Source",],cd["Target",],cd["interactions",])
dimnames(x)
dim(x)[3]
a = x[,,1]
for (i in 1:dim(x)[3]){
  t = x[,,i]
  a[t == 1] = dimnames(x)[[3]][i]
}
```


```{r}
library(pheatmap)

b = as.matrix.data.frame(a,rownames.force = T)
colnames(b) = colnames(a)
b = apply(b,MARGIN = c(1,2), function(x)as.numeric(x))

pheatmap(b, display_numbers = T,scale = "none",cluster_cols = F,cluster_rows = F )
setEPS()
postscript("heatmap_cellphonedb.eps")
pheatmap(b, display_numbers = T,scale = "none",cluster_cols = F,cluster_rows = F,color = c("lightgrey","lightyellow","gold","orange","darkorange","red","darkred" ))
dev.off()
```

```{r}
#expressed by opc-2 but not oligos, target mg 4 but not mg 5
cellphonedb[!is.na(cellphonedb$opcs_2.microglia_4) & is.na(cellphonedb$opcs_2.microglia_5),c("interacting_pair","secreted","opcs_2.microglia_4")]
#

#expressed by opc-9 dividing but not oligos, target mg4 but not 5
cellphonedb[!is.na(cellphonedb$opcs_9.microglia_4) & is.na(cellphonedb$opcs_9.microglia_5),c("interacting_pair","secreted","opcs_9.microglia_4")]

#expressed by mg4 but not 5, target opc2
cellphonedb[!is.na(cellphonedb$microglia_4.opcs_2) & is.na(cellphonedb$microglia_5.opcs_2),c("interacting_pair","secreted","microglia_4.opcs_2")]

#expressed by mg4 but not 5, target opc9 dividing
cellphonedb[!is.na(cellphonedb$microglia_4.opcs_2) & is.na(cellphonedb$microglia_5.opcs_9),c("interacting_pair","secreted","microglia_4.opcs_9")]

```

