---
title: "Subset_script"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r}
#install.packages("BiocManager")

#BiocManager::install("Seurat")
#BiocManager::install("ggplot2")
#BiocManager::install("sctransform")

library(Seurat)
library(ggplot2)
library(sctransform)
library(dplyr)
library(ape)
library(variancePartition)
library(cowplot)
library(Matrix)
library(magrittr)

#find out where you are
getwd()

#Specify where your matrix files are
dir= "/Users/whippoorwill/Desktop/Sequencing/LD_AVM02/"
```

```{r}
filename = "sobject.RData"
load(file = file.path(dir,"Data/Seurat",filename))
```


Subset the data to include only your celltypes of interest: 
```{r}
#edit keep to include all wanted celltypes
keep = c("mpeg")

#subset the object by the metadata column "celltype"
sobject = subset(sobject,subset = celltype %in% keep)

#check the resulting subset
table(sobject$celltype,sobject$seurat_clusters)
```
Whenever you remove cells from a seurat object, you have to re-normalize, scale, and calculate pca/umap/clustering. This is copied out from above. I'm keeping the same add.genes but of course you could change them here. 

```{r}
sct = FALSE
add.genes = c("ptprc","lgmn","hexb","mpeg1.1","apoeb")
regress = c("nFeature_RNA")

if (sct){sobject <- SCTransform(sobject, vars.to.regress = regress, verbose = FALSE,variable.features.n = 6000,conserve.memory = T)}

if (!sct){
  sobject <- NormalizeData(sobject,normalization.method = "LogNormalize", scale.factor = 10000)

  sobject<-FindVariableFeatures(sobject, selection.method = "vst", nfeatures = 10000)
  all.genes<-rownames(sobject)
  var.genes = VariableFeatures(sobject)

  add.genes = add.genes[!add.genes %in% var.genes]
  any(add.genes %in% var.genes)
  scalegenes = c(var.genes,add.genes)

  sobject<-ScaleData(sobject,features = scalegenes, vars.to.regress = regress)
}

top10 <- c(head(VariableFeatures(sobject),10),"cd74b","mpeg1.1","cd74a","hexb","apoeb")
plot1 = VariableFeaturePlot(sobject)
LabelPoints(plot=plot1, points = top10, repel = F, xnudge = 0.1, ynudge = 0.5)
```

Run PCA analysis
```{r}
sobject <- RunPCA(sobject,features = scalegenes,npcs = 50, verbose = FALSE)

ElbowPlot(sobject,ndims = 50, reduction = "pca")
print(sobject[["pca"]], dims = 1:10, nfeatures = 5)
```

Once you are satisfied with pc's, run clustering: 
```{r}
pcs = 1:30
sobject<-RunUMAP(sobject,reduction = "pca",dims = pcs, verbose = F)
sobject<-FindNeighbors(sobject,dims=pcs,verbose=F)

res = 0.3 #set this to anything
sobject<-FindClusters(sobject,verbose=F,resolution = res) 
```

Look at resulting clustering to evaluate
```{r}
genes = c("mpeg1.1","apoeb","ptprc","hexb","cd74a","cd74b","bzw2","g0s2","ctsba","lygl1")
genes[genes %in% VariableFeatures(sobject)]
dims = c("seurat_clusters","sample_description","age")
QC = c("nCount_RNA","nFeature_RNA","percent.mito","percent.ribo")

for (dim in dims){
  print(DimPlot(sobject,group.by = dim, label = T))
}

FeaturePlot(sobject,genes)
FeaturePlot(sobject,QC)

VlnPlot(sobject,genes,sort = "increasing",pt.size = 0.01)

#Build a clustering tree
Idents(sobject) = "seurat_clusters"
sobject= BuildClusterTree(sobject,dims = 1:30)
tree = sobject@tools$BuildClusterTree
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
save(sobject,file = file.path(dir,paste0(Project(sobject),"_",keep,"only_umap.RData")))
```

Heatmap for clusters within this celltype
```{r}
Idents(sobject) = "seurat_clusters"
markers_all <- FindAllMarkers(
    object = sobject,
    features = rownames(sobject),
    test.use = "MAST",
    only.pos = FALSE, 
    min.pct = 0.05, #gene must be present in 5% of the cells in the cluster
    logfc.threshold = 0.0
)

write.csv(markers_all,file = file.path(dir,paste0(keep,"_all_markers.csv")))
```


```{r}
#set your p-value threshold
pcut = 1e-5
genespercluster = 5

#select positive markers for heatmap
markers_all = markers_all[markers_all$avg_logFC>0.2,]

#select genes that pass a p-value threshold
markers_all = markers_all[markers_all$p_val_adj<pcut,]
markers_all <- markers_all[markers_all$gene %in% names(table(markers_all$gene))[table(markers_all$gene) == 1],] #markers that define a single cluster

topgenes <- markers_all %>% group_by(cluster) %>% top_n(genespercluster, avg_logFC)

setEPS()
postscript(file.path(dir,paste0("heatmap_",keep,"_only",pcut,".eps")))
DoHeatmap(
    object = sobject, 
    features = topgenes$gene,
    group.by = 'seurat_clusters',
    size = 5,
    label = T,
    draw.lines = T
)

dev.off()
```
Heatmap for console

```{r}
DoHeatmap(
    object = sobject, 
    features = topgenes$gene,
    group.by = 'seurat_clusters',
    size = 5,
    label = T,
    draw.lines = T
)
```


```{r}
save(sobject,file = file.path(dir,paste0(keep,"_only.RData")))
```

As this is a checkpoint, you can load in the data again if necessary: 

```{r}
load(file.path(dir,"mpeg_only_sctransform.RData"))
de = read.csv(file.path(dir,"all_markers_mpeg.csv"),stringsAsFactors = F)
head(de)
```

```{r}
genes = c("pcna","siglec15l")
DimPlot(sobject,label = T)
DimPlot(sobject,group.by = "sample_description")

FeaturePlot(sobject,genes)
VlnPlot(sobject,genes,pt.size = 0.001,sort = T)

sobject = BuildClusterTree(sobject,dims = 1:30)
PlotClusterTree(sobject)
```


```{r}
table(sobject$age,sobject$seurat_clusters)
```


```{r}
type1 = c(0,2,5,8)
name1 = "juvenile"
type2 = c(1,4,7)
name2 = "adult"
type3 = c(3)
name3 = "macrophage"
type4 = 6
name4 = "proliferating"

#Initialize the cluster levels as a vector and replace the cluster levels with the appropriate name. 
clusters = as.factor(sobject$seurat_clusters)
type = levels(clusters)
type[type1+1] = name1
type[type2+1] = name2
type[type3+1] = name3
type[type4+1] = name4
levels(clusters) = type

#Add a metadata column
sobject$celltype = clusters
sobject$celltypecluster = paste0(sobject$celltype,"-",sobject$seurat_clusters)

DimPlot(sobject,group.by = "celltypecluster",split.by = "age")
```

Block to print multiple graphs: 
```{r}
name = "all_mpeg"
genes = c("marco","mpeg1.1","apoeb","p2ry12","ptprc","hexb","irf8","spi1b","spi1a","csf1ra","csf1rb","slc7a7","c1qa","c1qb","c1qc","cd74a","cd74b","pcna","mki67","siglec15l")
features = c("percent.mito","percent.ribo","nCount_RNA","nFeature_RNA")
groups = c("age","sample_description","seurat_clusters","celltype","celltypeclusters")
genes = genes[genes %in% rownames(GetAssayData(sobject,slot = "data"))]

#just to print to the console some nice graphs
FeaturePlot(sobject,genes[1:5])
FeaturePlot(sobject,genes[6:9])

#feature plots
for(feature in c(genes,features)){
  PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
}

#split feature plots by individual
for(feature in c(features)){
  PrintSeuratGraph(namecard = paste0(name,"_split"),graphtype = "feature",feature = feature,split = "sample_description")
}

#dim plots for clustering
for(group in groups){
  PrintSeuratGraph(namecard = name,graphtype = "dim",group = group, feature = group)
}

#violin plots
for(feature in c(genes,features)){
  PrintSeuratGraph(namecard = name,graphtype = "violin",feature = feature,group = "seurat_clusters")
}
```

```{r}
Idents(sobject) = "celltypecluster"
sobject = BuildClusterTree(sobject,dims = c(1:30))
tree = sobject@tools$BuildClusterTree
setEPS()
postscript(file.path(dir,paste0(name,"tree.eps")))
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
dev.off()
```

Conduct differential expression analysis on the new metadata
```{r}
Idents(sobject) = "celltype"
markers_all = FindAllMarkers(
  object = sobject,
  features = rownames(sobject),
  test.use = "MAST", 
  only.pos = FALSE, 
  min.pct = 0.10, 
  logfc.threshold = 0.0)

```

```{r}
filename = "juv_v_adult_mpeg"
write.csv(markers_all,file = file.path(dir, paste0(filename,".csv")))

single = TRUE
pcut = 1e-25
lfc = 0.2
ngenes = 5

markers_all = markers_all[markers_all$p_val_adj<pcut,]
markers_all = markers_all[markers_all$avg_logFC>lfc,]

#markers that define a single cluster
if (single){
  markers_all_single = markers_all[
    markers_all$gene %in% names(table(markers_all$gene))[table(markers_all$gene) == 1],]
} 


topgenes <- markers_all %>% group_by(cluster) %>% top_n(ngenes, avg_logFC)

setEPS()
postscript(file.path(dir,paste0("heatmap_",filename,".eps")))
DoHeatmap(
    object = sobject, 
    features = topgenes$gene,
    group.by = 'celltypecluster',
    size = 5,
    label = T,
    draw.lines = T
)

dev.off()

```

Heatmap for console
```{r}
DoHeatmap(
    object = sobject, 
    features = topgenes$gene,
    group.by = 'celltypecluster',
    size = 5,
    label = T,
    draw.lines = T
)
```

