---
title: "Seurat_Code_PostQC_2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown


```{r}
#install.packages("BiocManager")

#BiocManager::install("Seurat")
#BiocManager::install("ggplot2")
#BiocManager::install("sctransform")

library(Seurat)
library(ggplot2)
library(sctransform)
library(dplyr)
library(ape)
library(variancePartition)
library(cowplot)
library(Matrix)
library(org.Mm.eg.db)
library(org.Hs.eg.db)
library(org.Dr.eg.db)

#find out where you are
getwd()
dir= "/Users/whippoorwill/Desktop/Sequencing/Nick/2_2020"
 #set this to whatever you like, above the level of your data folder
#setwd(dir)
```
Function to print multiple graphs: 
```{r}
PrintSeuratGraph = function(namecard = "a",seurat_object = sobject,graphtype = "feature",feature = NULL,group = NULL,split=NULL,cellnames=NULL){
  if (!is.null(cellnames)){
    Idents(seurat_object) = cellnames[1]
    cells = colnames(seurat_object)[Idents(seurat_object) %in% cellnames[2:length(cellnames)]]} 
  else {cells = cellnames}
  if (graphtype == "feature"){
    graph = FeaturePlot(seurat_object,features = feature,split.by = split, cells = cells,cols = c("lightgrey","lightblue","yellow","orange","darkorange","red"))
  }
  if (graphtype == "violin"){
    graph = VlnPlot(seurat_object,features = feature, pt.size = 0.1, idents = cellnames[2:length(cellnames)],group.by = group, split.by = split)
  }
  if (graphtype == "dim"){
    graph = DimPlot(seurat_object,cells = cells, group.by = group, split.by = split)
    
  }
  name = paste0(feature,"_",graphtype,namecard,".eps")
  graph
  setEPS()
  postscript(file.path(dir,"plots",name))
  print(graph)
  dev.off()
}

```

```{r}
load(file.path(dir,"NickData/Nick2020_unfiltered.RData"))
```

Set limits based on the graphs from the QC page
```{r}
mlo = -0.1
mhi = 10
clo = 1200
chi = 15000
flo = 500
fhi = 3000
rlo = 0
rhi = 70
```

Subset the data. Printing tables lets you see the progression as you tighten your QC cutoffs. 
```{r}
print("initial")
table(sobject$sample_description) #print out initial cell counts
sobject = subset(sobject, subset = percent.mito>mlo & percent.mito < mhi) #subset mitochondria
print("mito")
table(sobject$sample_description) #new cell counts
sobject = subset(sobject, subset = nCount_RNA>clo & nCount_RNA < chi)
print("ncount")
table(sobject$sample_description)
sobject = subset(sobject, subset = nFeature_RNA>flo & nFeature_RNA < chi)
print("nfeature")
table(sobject$sample_description)
```

```{r}
save(sobject, file = file.path(dir,"sobject_filtered_umap.RData"))
```

Find variable features, normalize, scale, run PCA, clustering, umap

Include any genes that may not vary a lot but are important for downstream applications.PSA: mouse genes have the first letter capitalized; humans are fully capitalized; fish are lowercase. rownames(sobject) will help you troubleshoot. 
```{r}
add.genes = c('ptprc','lgmn','hexb','apoeb','mpeg1.1')
regress = c("nFeature_RNA")
sct = FALSE
```

The following is the standard method of normalization and scaling. Interchangeable with the next chunk. 
```{r}
if (!sct){
  sobject <- NormalizeData(sobject,normalization.method = "LogNormalize", scale.factor = 10000)

  sobject<-FindVariableFeatures(sobject, selection.method = "vst", nfeatures = 10000)
  all.genes<-rownames(sobject)
  var.genes = VariableFeatures(sobject)

  labels <- c(head(VariableFeatures(sobject),10),"cd74b","mpeg1.1","cd74a","hexb","apoeb")
  plot1 = VariableFeaturePlot(sobject)
  LabelPoints(plot=plot1, points = labels, repel = F, xnudge = 0.1, ynudge = 0.5)

  add.genes = add.genes[!add.genes %in% var.genes]
  any(add.genes %in% var.genes)
  scalegenes = c(var.genes,add.genes)

  sobject<-ScaleData(sobject,features = scalegenes, vars.to.regress = regress)
}

```

Alternative: SCTransform (great for smaller datasets)
```{r}
if (sct){
  add.genes = c("ptprc","lgmn","hexb","mpeg1.1","apoeb")
  regress = c("nFeature_RNA","percent.mito")

  sobject <- SCTransform(sobject, vars.to.regress = regress, verbose = FALSE,variable.features.n = 6000,conserve.memory = T)


  top10 <- c(head(VariableFeatures(sobject),10),"cd74b","mpeg1.1","cd74a","hexb","apoeb")
  top10 <- head(VariableFeatures(sobject),10)
  plot1 = VariableFeaturePlot(sobject)
  LabelPoints(plot=plot1, points = top10, repel = F, xnudge = 0.1, ynudge = 0.5)
}

```

Run PCA analysis
```{r}
sobject <- RunPCA(sobject,features = scalegenes,npcs = 50, verbose = FALSE)

ElbowPlot(sobject,ndims = 50, reduction = "pca")
print(sobject[["pca"]], dims = 1:20, nfeatures = 5)
```

Once you are satisfied with pc's, run clustering: 
```{r}
pcs = 1:30
sobject<-RunUMAP(sobject,reduction = "pca",dims = pcs, verbose = F)
sobject<-FindNeighbors(sobject,dims=pcs,verbose=F)

res = 0.8 #set this to anything
sobject<-FindClusters(sobject,verbose=F,resolution = res) 
```

Plot important objects; check parameters before moving forward, evaluate QC, clustering
```{r}
genes = c("mpeg1.1","apoeb","ptprc","hexb")
dims = c("seurat_clusters","sample_description","age")
QC = c("nCount_RNA","nFeature_RNA","percent.mito","percent.ribo")

for (dim in dims){
  print(DimPlot(sobject,group.by = dim, label = T))
}

FeaturePlot(sobject,genes)
FeaturePlot(sobject,QC)

VlnPlot(sobject,genes,sort = "increasing",pt.size = 0.01)

#Build a clustering tree
Idents(sobject) = "seurat_clusters"
sobject= BuildClusterTree(sobject,dims = 1:30)
tree = sobject@tools$BuildClusterTree
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
```
```{r}
#If you want to see a certain cluster: 
cluster = '13'

#select only the cells from that cluster
cells = sobject$seurat_clusters
cells = cells[cells == cluster]

DimPlot(sobject,cells.highlight = names(cells))
```

If you are happy with the QC, you can move on to either subdividing further to isolate interesting cells (below) or running differential expression comparisons on the existing clustering. 

Note that the chunk below requires you to manually decide which clusters to keep or exclude. The violin plots will help decide based on marker genes. You can also choose to annotype (as many types as you want) and skip subsetting. 
```{r}
#Edit this part carefully. You can add any number of types. Each cluster can only be one type. 
type1 = c(0,1,3,5,6,7,8,9,10,15,18,19,20)
name1 = "mpeg"
type2 = c(2,4,11,12,13,14,16,17,21,22,23) 
name2 = "cd45"

#Initialize the cluster levels as a vector and replace the cluster levels with the appropriate name. 
clusters = as.factor(sobject$seurat_clusters)
type = levels(clusters)
type[type1+1] = name1
type[type2+1] = name2
levels(clusters) = type

#Add a metadata column
sobject$celltype = clusters

#check the celltype assignment for accuracy
table(sobject$celltype,sobject$seurat_clusters)

#Check them against your marker genes
markergene = c("mpeg1.1","hexb","ptprc")
VlnPlot(sobject,markergene,group.by = "celltype",pt.size = 0.01)

#add a metadata column labelling each cluster
sobject$celltypecluster = paste0(sobject$celltype,"-",sobject$seurat_clusters)

#save the unsubsetted data. This will overwrite the one you just saved because it has all the same info + celltype assigned and tree calculated. 
save(sobject,file = file.path(dir,paste0(Project(sobject),"_sobject_filtered_umap.RData")))
```

Save the tree
```{r}
Idents(sobject) = "celltypecluster"
sobject= BuildClusterTree(sobject,dims = 1:30)
tree = sobject@tools$BuildClusterTree
setEPS()
postscript(file.path(dir,"tree_iteration1.eps"))
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
dev.off()
```

Block to print multiple graphs: 
```{r}
name = "juv_mpeg"
genes = c("marco","mpeg1.1","apoeb","p2ry12","ptprc","hexb","irf8","spi1b","spi1a","csf1ra","csf1rb","slc7a7","c1qa","c1qb","c1qc","cd74a","cd74b","pcna","mki67","siglec15l")
features = c("percent.mito","percent.ribo","nCount_RNA","nFeature_RNA")
groups = c("age","sample_description","seurat_clusters","celltype","celltypeclusters")
genes = genes[genes %in% rownames(GetAssayData(sobject,slot = "data"))]

#just to print to the console some nice graphs
FeaturePlot(sobject,genes[1:5])
FeaturePlot(sobject,genes[6:9])

#feature plots

for(feature in genes){
  PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
}

for(feature in features){
  PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
}

#split feature plots by individual
for(feature in c(features)){
  PrintSeuratGraph(namecard = paste0(name,"_split"),graphtype = "feature",feature = feature,split = "sample_description")
}

#dim plots for clustering
for(group in groups){
  PrintSeuratGraph(namecard = name,graphtype = "dim",group = group, feature = group)
}

#violin plots
for(feature in c(genes,features)){
  PrintSeuratGraph(namecard = name,graphtype = "violin",feature = feature,group = "seurat_clusters")
}
```


