---
title: "Atac-seq"
output: html_notebook
---
```{r}
library(Seurat)
library(dplyr)
library(ggplot2)
```


Load file with differentially expressed genes

```{r}
#if public: 
#genes = read.csv("https://github.com/lcdorman/scrnaseq/blob/master/allmarkers_vargenesMG_0501.csv",stringsAsFactors = F)

#if private:
genes = read.csv("allmarkers_vargenesMG_0501.csv",stringsAsFactors = F)

#remove excess columns
genes = genes[,3:ncol(genes)]
head(genes)

#remove genes that don't pass a threshold for p-value and log fold change
alpha = 1e-25
lfoldchange = 0.2

genes = genes[genes$p_val_adj<alpha,]
genes = genes[genes$avg_logFC > lfoldchange | genes$avg_logFC < -lfoldchange,]
```

Load file with atac-seq peaks and edit to merge each gene into one line

```{r}
#if public: 
#peaks = read.csv("https://github.com/lcdorman/scrnaseq/blob/master/MergedpeaksPBSAnnotated_v2.csv",stringsAsFactors = F)

#if private:
peaks = read.csv("MergedpeaksPBSAnnotated_v2.csv",stringsAsFactors = F)
head(peaks)
peaks = peaks[,c("Peak.Score","Focus.Ratio.Region.Size","Detailed.Annotation","Gene.Name")]
head(peaks)
```

Change "Detailed.Annotation" to specific meaningful values
```{r}
annotation = peaks$Detailed.Annotation
rpt = grep("repeat",annotation)
names(rpt) = rep("Repeat",length(rpt))

#pull out different kinds of annotations
prom = grep("promoter",annotation)
names(prom) = rep("Promoter",length(prom))

intergenic = grep("Intergenic",annotation)
names(intergenic) = rep("Intergenic",length(intergenic))

cpg = grep("CpG",annotation)
names(cpg) = rep("CpG",length(cpg))

utr3 = grep("3' UTR",annotation)
names(utr3) = rep("UTR3",length(utr3))

utr5 = grep("5' UTR",annotation)
names(utr5) = rep("UTR5",length(utr5))

exon = grep("exon",annotation)
names(exon) = rep("Exon",length(exon))
exon = exon[!exon %in% c(utr3,utr5)]

intron = grep("intron",annotation)
names(intron) = rep("Intron",length(intron))

tts = grep("TTS",annotation)
names(tts) = rep("TTS",length(tts))

annotationfull = c(rpt,exon,prom,intergenic,cpg,utr3,utr5,intron,tts)
annotationfull = annotationfull[order(annotationfull)]

annotation[annotationfull] = names(annotationfull)
```

Annotate peaks
```{r}
peaks$Detailed.Annotation = annotation

#Separate only the genes that are present in my file
allgenes = genes$gene
peaks = peaks[peaks$Gene.Name %in% allgenes,]

#Include only confident peaks (> 1 sample)
peaks = peaks[grep("\\|",peaks$Focus.Ratio.Region.Size),]
head(peaks)
```

Select only the genes with promoter peaks (skip or change "Promoter" based on what you are interested in looking at)
```{r}
peaks = peaks[peaks$Detailed.Annotation == "Promoter",]
peaks = peaks[order(peaks$Peak.Score,decreasing = T),]
head(peaks)
```

Merge every annotation for the same gene into one line. 
```{r}
k = ncol(peaks)
for (i in 2:length(peaks$Gene.Name)){
  if (peaks$Gene.Name[(i-1)] == peaks$Gene.Name[i]){ 
    peaks[(i-(k/4)),(k+1):(k+4)] = peaks[i,1:4]
    k = k+4
  } else {k = 4}
}

peaks = peaks[!duplicated(peaks$Gene.Name),]
head(peaks)
```

Select only the annotations for genes in your file, and save for future reference
```{r}
peaks = peaks[match(genes$gene,peaks$Gene.Name),]
atac =cbind(genes,peaks)
write.csv(atac,file = "atacseq_degenes_promoters.csv")
```

(If starting in the middle, load the previously saved file)
```{r}
atac = read.csv("atacseq_degenes_promoters.csv",stringsAsFactors = F)
```

Make a new spreadsheet with only the genes upregulated in a specific cluster
```{r}
cluster = "3"
genes_cluster = genes[genes$cluster == cluster,]
dim(genes_cluster)
peaks_cluster = peaks[peaks$Gene.Name %in% genes_cluster$gene,]
dim(peaks_cluster)
head(peaks_cluster)
```

Write a spreadsheet (note: 315 genes in cluster 3, 136 peaks)

```{r}
write.csv(peaks_cluster,paste0("Peaks_cluster",cluster,".csv"))
```

Compare this to, say, cluster 4: 

Make a new spreadsheet with only the genes upregulated in cluster 4

```{r}
cluster = "4"
genes_cluster = genes[genes$cluster == cluster,]
peaks_cluster = peaks[peaks$Gene.Name %in% genes_cluster$gene,]
write.csv(peaks_cluster,paste0("Peaks_cluster",cluster,".csv"))
```


Compare this to, say, cluster 8: 

Make a new spreadsheet with only the genes upregulated in cluster 8
```{r}
cluster = "8"
genes_cluster = genes[genes$cluster == cluster,]
peaks_cluster = peaks[peaks$Gene.Name %in% genes_cluster$gene,]
write.csv(peaks_cluster,paste0("Peaks_cluster",cluster,".csv"))
```

Compare with barres dataset: (Courtesy of http://brainrnaseq.org)

```{r}
#private
barres = read.csv("barreslab_rnaseq.csv",stringsAsFactors = F)

#public
#barres = read.csv("https://github.com/lcdorman/scrnaseq/blob/master/barreslab_rnaseq.csv",stringsAsFactors = F)

```

Calculate a cell type specificity rating for your cell type of interest (in this case, microglia)

```{r}
head(barres)
colnames(barres)
```
Specify which columns are your cell of interest (coi) and which are other cells (other)

```{r}
coi = 8
other = c(3:(coi-1),(coi+1):9)
```

This function will calculate a Cell Type specificity by dividing the fpkm for your cell type by the average fpkm for all other cell types.
```{r}
for (i in 1:nrow(barres)){
  row = barres[i,]
  mspec = as.numeric(row[coi])/mean(as.numeric(row[other]),na.rm = T)
  barres[i,10] = mspec
}
colnames(barres)[10] = "Specificity"
```


Add the barres annotation to the atac-seq gene datasheet
```{r}
barres = barres[match(atac$gene,barres$Gene.symbol),]
dim(atac)
dim(barres)
atac = cbind(barres,atac)
atac$Gene.symbol = atac$gene
head(atac)
```
Save the annotated file

```{r}
write.csv(atac,file = paste0("atacseqwithbarres_promoters_",colnames(barres)[coi],"_specificity.csv"))
```


Relationship between peak size and microglial specificity by cluster
```{r}
a = read.csv(file = paste0("atacseqwithbarres_promoters_",colnames(barres)[coi],"_specificity.csv"),stringsAsFactors = F)
#a = cbind("Gene" = atac$Gene.symbol,"Cluster" = atac$cluster,"Specificity" = atac$Specificity,"Peak" = atac$Peak.Score)

a$Peak.Score = as.character(a$Peak.Score)
a$Peak.Score = as.numeric(a$Peak.Score)

a$Specificity = as.character(a$Specificity)
a$Specificity = as.numeric(a$Specificity)

a$cluster = as.character(a$cluster)
head(a)
```

adding in spliced:unspliced ratio calculated in scvelo (see scvelo_mgavm02 script for details)

```{r}
#ratio = read.csv("https://github.com/lcdorman/scrnaseq/blob/master/spliced_unspliced_ratio.csv",stringsAsFactors = F,header = F)
#genes = read.csv("https://github.com/lcdorman/scrnaseq/blob/master/splicegenes.csv",stringsAsFactors = F)
ratio = read.csv("spliced_unspliced_ratio.csv",stringsAsFactors = F,header = F)
ratio = as.numeric(ratio)
genes = read.csv("splicegenes.csv",stringsAsFactors = F)
names(ratio) = genes$Name
```

add ratios to the spreadsheet a
```{r}
ratio = ratio[match(a$gene,names(ratio))]
a$splicedratio = ratio
```

Peak chromatin score vs splicedratio
```{r}
ac = a[a$cluster %in% clusters & a$avg_logFC>0,]
ac = ac[!is.na(ac$Peak.Score),]
ac = ac[!is.na(ac$splicedratio),]
ac$splicedratio[ac$splicedratio == "Inf"] =5000

b <- ggplot(ac, aes(x = log(splicedratio), y = Peak.Score))


c = b + geom_point(aes(color = cluster)) +
  scale_color_manual(values = c('green','blue','darkgrey','purple')) + geom_text( 
    data=ac %>% filter(log(splicedratio)>6 | log(splicedratio)<(-3) | Peak.Score>120),
    aes(label=gene),nudge_x = -0.8,check_overlap = T)   
c
```

Print a graph
```{r}
setEPS()
postscript("atacseqpeak_vs_splicedratio.eps")
c
dev.off()
```

Peak chromatin score vs specificity
```{r}
b <- ggplot(ac, aes(y = Peak.Score, x = log(Specificity)))


c = b + geom_point(aes(color = cluster)) +
  scale_color_manual(values = c('green','blue','darkgrey','purple')) + geom_text( 
    data=ac %>% filter(log(Specificity)>5 | log(Specificity)<(-3) | Peak.Score>100),aes(label=gene),nudge_x = -0.5,check_overlap = T)
c
```


```{r}
setEPS()
postscript("~/Desktop/specificity_peak.eps")
c
dev.off()
```

Specificity vs spliced:unspliced ratio

```{r}
b <- ggplot(ac, aes(x = log(splicedratio), y = log(Specificity)))


c = b + geom_point(aes(color = cluster)) +
  scale_color_manual(values = c('green','blue','darkgrey','purple')) + geom_text( 
    data=ac %>% filter(log(splicedratio)>6 | log(splicedratio)<(-3) | log(Specificity)>5),
    aes(label=gene),nudge_x = -0.8,check_overlap = T)   
c
```


```{r}
setEPS()
postscript("~/Desktop/specificity_splicedratio.eps")
c
dev.off()
```


Volcano Plot: 
```{r}
#fc = markers_all
de = read.csv("allmarkers_vargenesMG_0501.csv",stringsAsFactors = F)
colnames(fc)[8] = "Gene"
```

```{r}
cluster = 3
alpha = 1e-25
foldchange = 1.15


newlist = list()
#Split by cluster
for (i in c(1:9)){
  newlist[[i]] = de[de$cluster == i,]
}

#select a single cluster
fc = newlist[[cluster]]
fc = fc[!is.na(fc$avg_logFC),]
colorkeysdown = fc$Gene[fc$avg_logFC < -log2(foldchange) & fc$p_val_adj < alpha]
colorkeysup = fc$Gene[fc$avg_logFC > log2(foldchange) & fc$p_val_adj < alpha]

xlim = max(abs(fc$avg_logFC))
```


```{r}
atac = atac[atac$Peak.Score>5,]

allcolors = rep("darkgrey",length(fc$Gene))
names(allcolors) = fc$Gene

allcolors[names(allcolors) %in% colorkeysdown] = "blue"
allcolors[names(allcolors) %in% colorkeysup]= "yellow"
allcolors[names(allcolors)%in% atac$gene] = "red"


names(allcolors)[allcolors == "yellow"] = "u"
names(allcolors)[allcolors == "darkgrey"] = "-"
names(allcolors)[allcolors == "blue"] = "d"
names(allcolors)[allcolors == "red"] = "a"

```

Make the plot
  
```{r}
library(EnhancedVolcano)
setEPS()
postscript(paste0("volcano_atac",cluster,"label.eps"))
EnhancedVolcano(fc,
                lab = fc$Gene,
                x = 'avg_logFC',
                y = 'p_val_adj',
                xlim = c(-xlim,xlim),
                title = paste0("atac",cluster),
                subtitle = "",
                drawConnectors = F,
                legendPosition = 'right',
                legendVisible = F,
                pCutoff = alpha,
                FCcutoff = log2(foldchange),
                selectLab = c(colorkeysdown,colorkeysup),
                transcriptPointSize = 3,
                transcriptLabSize = 5,
                col=c('black', 'black', 'black', 'red3'),
                colCustom = allcolors,
                gridlines.major = F,
                gridlines.minor = F,
                colAlpha = 1)

dev.off()
```
Make a table

Select only the diff. exp genes from atac-seq data
```{r}
atac$cluster = as.factor(atac$cluster)
atac = atac[atac$p_val_adj < alpha,]
atacup = atac[atac$avg_logFC > log2(foldchange),]
atacdown = atac[atac$avg_logFC < -log2(foldchange),]

atacup = table(atacup$cluster)
atacdown = table(atacdown$cluster)
```
compare to atac-seq data
```{r}
deup = de[de$avg_logFC>log2(foldchange) & de$p_val_adj <alpha,]
dedown = de[de$avg_logFC< -log2(foldchange) & de$p_val_adj <alpha,]

allup= table(deup$cluster)
alldown = table(dedown$cluster)

t = rbind(allup,atacup,pctup = round(atacup/allup*100,0),alldown,atacdown, pctdown = round(atacdown/alldown*100,0))
```

Bar chart

```{r}
#pull out exactly what you want to plot
p = t[c("pctdown","pctup"),]
c = barplot(p, horiz = F, las=1, xlim = c(0,30),xlab = "cluster", ylab = '% promoter atac-seq peaks',
        beside=T, col=c('blue','red'),ylim = c(0,100),legend = c("downregulated","upregulated"),axis.lty = 1)
c
#actually make a bargraph of t$rup and t$rdown
setEPS()
postscript("barplot_atacseq_updown.eps")
c
dev.off()
```


Cellphone db: also from jupyter, see python script for details. "X3.04" means from cluster 3 (source) to cluster 4 (target)
```{r}
cellphonedb = read.csv("significant_means.csv",stringsAsFactors = F)
#cellphonedb = read.csv("https://github.com/lcdorman/scrnaseq/blob/master/significant_means.csv",stringsAsFactors = F)
colnames(cellphonedb)
```
count per interaction, remove rows with nothing
```{r}
cdb = cellphonedb[,13:28]
cdb = cdb[rowSums(cdb,na.rm = T)>0,]
```

Replace NA with 0 and any number with 1 (count incidents)
```{r}
cdb[!is.na(cdb)] = 1
cdb[is.na(cdb)] = 0
cd = colSums(cdb,na.rm = T)
cd
```
**Make a table(* 

First we will repeat the names for source and target and then rename source and target to the appropriate portions of the name itself. 

```{r}
cd = rbind("Source" = names(cd),"Target" = names(cd),"interactions" = cd)
cd["Source",] = sapply(cd["Source",],function(x)paste0("Source_",strsplit(x,"")[[1]][2]))
cd["Target",] = sapply(cd["Target",],function(x)strsplit(x,"\\.")[[1]][2]) #split on ".", take 2nd
cd["Target",] =sapply(cd["Target",],function(x)paste0("Target_",strsplit(x,"")[[1]][2])) #remove "0"
```

Make a 3D table first, then iterate through each matrix and add the 3rd dimension (#interactions) to its appropriate position

```{r}
x = table(cd["Source",],cd["Target",],cd["interactions",])
dimnames(x)
dim(x)[3]
a = x[,,1]
for (i in 1:dim(x)[3]){
  t = x[,,i]
  a[t == 1] = dimnames(x)[[3]][i]
}
```


```{r}
library(pheatmap)

b = as.matrix.data.frame(a,rownames.force = T)
colnames(b) = colnames(a)
b = apply(b,MARGIN = c(1,2), function(x)as.numeric(x))

pheatmap(b, display_numbers = T,scale = "none",cluster_cols = F,cluster_rows = F )
setEPS()
postscript("heatmap_cellphonedb.eps")
pheatmap(b, display_numbers = T,scale = "none",cluster_cols = F,cluster_rows = F,color = c("lightgrey","lightyellow","gold","orange","darkorange","red","darkred" ))
dev.off()
```






